utils_bsd.go

```diff
@@ -0,0 +1,23 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+//go:build darwin || dragonfly || freebsd || netbsd || openbsd
+
+package kmsg
+
+import (
+	"fmt"
+	"time"
+
+	"golang.org/x/sys/unix"
+)
+
+func getBootTime() (time.Time, error) {
+	timeval, err := unix.SysctlTimeval("kern.boottime")
+	if err != nil {
+		return time.Time{}, fmt.Errorf("could not get boot time: %w", err)
+	}
+
+	return time.Now().Add(-1 * time.Since(time.Unix(timeval.Unix()))), nil
```

Comment:
if `kern.boottime` gives a timestamp, it can be converted directly to `time.Time` without `time.Since`/`time.Now`

https://pkg.go.dev/time#Unix

Reply:
AFAIU we can't get the timestamp right away, we first have to deal with an unsafe pointer. Check out this implementation: https://github.com/mackerelio/go-osstat/blob/v0.2.4/uptime/uptime_bsd.go which finally also resorts to a Timeval.
So using [SysctlTimeval](https://pkg.go.dev/golang.org/x/sys@v0.6.0/unix?GOOS=darwin#SysctlTimeval) makes it a bit more straightforward.

Reply:
my point is that: `time.Since(ts) == now() - ts`, so what you're doing is `now() - (now() - ts) == ts`

you need `time.Unix(timeval.Unix())`

Reply:
Sorry I got confused. Fixed.